stdlib/mexpr/type-check.mc

```diff
@@ -0,0 +1,436 @@
+  | TyAll t ->
+    match genBase lvl t.ty with (vars, ty) then
+      (vars, TyAll {t with ty = ty})
+    else never
+end
+
+lang ConstTypeGeneralize = Generalize + ConstTypeAst
+  sem instBase (subst : Map Name Type) =
+  | (TyUnknown _ | TyInt _ | TyBool _ | TyFloat _ | TyChar _) & ty -> ty
+  sem genBase (lvl : Level) =
+  | (TyUnknown _ | TyInt _ | TyBool _ | TyFloat _ | TyChar _) & ty -> ([], ty)
+end
+
+lang SeqTypeGeneralize = Generalize + SeqTypeAst
+  sem instBase (subst : Map Name Type) =
+  | TySeq t ->
+    TySeq {t with ty = instBase subst t.ty}
+
+  sem genBase (lvl : Level) =
+  | TySeq t ->
+    match genBase lvl t.ty with (vars, ty) then
+      (vars, TySeq {t with ty = ty})
+    else never
+end
+
+-------------------
+-- TYPE CHECKING --
+-------------------
+
+lang TypeCheck = Unify + Generalize
+  -- Type check `tm', with FreezeML-style type inference.
+  sem typeCheck =
+  | tm -> typeCheckBase _tcEnvEmpty tm
+
+  sem typeCheckBase (env : TCEnv) =
+  | tm ->
+    let msg = join [
+      "Type check failed: type checking not supported for term\n",
+      use MExprPrettyPrint in expr2str tm
+    ] in
+    infoErrorExit (infoTm tm) msg
+end
+
+lang VarTypeCheck = TypeCheck + VarAst
+  sem typeCheckBase (env : TCEnv) =
+  | TmVar t ->
+    match _lookupVar t.ident env with Some ty then
+      let ty =
+        if t.frozen then ty
+        else inst env.currentLvl ty
+      in
+      TmVar {t with ty = ty}
+    else
+      let msg = join [
+        "Type check failed: reference to unbound variable\n",
+        "Var: ", nameGetStr t.ident, "\n"
+      ] in
+      infoErrorExit t.info msg
+end
+
+lang LamTypeCheck = TypeCheck + LamAst
+  sem typeCheckBase (env : TCEnv) =
+  | TmLam t ->
+    let tyX =
+      match t.tyIdent with TyUnknown _ then
+        -- No type annotation: assign a monomorphic type variable to x.
+        newvarWeak env.currentLvl
+      else
+        -- Type annotation: assign x its annotated type.
+        t.tyIdent
+    in
+    let body = typeCheckBase (_insertVar t.ident tyX env) t.body in
+    let tyLam = ityarrow_ t.info tyX (ty body) in
+    TmLam {{t with body = body}
+              with ty = tyLam}
+end
+
+lang AppTypeCheck = TypeCheck + AppAst
+  sem typeCheckBase (env : TCEnv) =
+  | TmApp t ->
+    let lhs = typeCheckBase env t.lhs in
+    let rhs = typeCheckBase env t.rhs in
+    let tyRes  = newvar env.currentLvl in
+    unify (ty lhs, tyarrow_ (ty rhs) tyRes);
+    TmApp {{{t with lhs = lhs}
+               with rhs = rhs}
+               with ty = tyRes}
+end
+
+lang LetTypeCheck = TypeCheck + LetAst
+  sem typeCheckBase (env : TCEnv) =
+  | TmLet t ->
+    let lvl = env.currentLvl in
+    let body = typeCheckBase {env with currentLvl = addi 1 lvl} t.body in
+    let inexpr =
+      match t.tyBody with TyUnknown _ then
+        -- No type annotation: generalize the inferred type
+        let genBody = gen lvl (ty body) in
+        typeCheckBase (_insertVar t.ident genBody env) t.inexpr
```

Comment:
Good catch. Yeah, updating `t.tyBody` seems sensible. Should I also update `t.tyIdent` for lambdas then, for consistency? That would make the pretty-printed AST display the types of all bindings after type-checking, which would be nice. The potential downside is that we forget whether a given annotation was provided by the programmer or if we added it afterwards, but I guess we don't really care about that.

Reply:
I implemented the approach above, but found a slight complication related to type variables. For instance, take the following snippet of code:
```
mexpr
lam x. let y = x in y
```
Typechecking the term and printing the result now produces the following:
```
lam x: _a.
  let y: _a =
    x
  in
  y
```
This cannot be parsed to produce the same term we started with, since it contains monomorphic type variables, for which there is no syntax. In addition, for the program
```
mexpr
let f = lam x. x in f
```
we get the result
```
let f: all a. (a) -> (a) =
  lam x: _a.
    x
in
f
```
where the parameter `x` is annotated with `_a` rather than `a`. I'm not quite sure what the best way of handling this is.

Reply:
I have reverted this change now as discussed in the meeting. Instead, we probably want a separate field for the type of the let-bound variable.